Switch to upstream gnu-efi (v4.0.4)

andreabolognani opened the pull request:
This is part of an attempt to bring riscv64 support into shim.

See #420 for the initial attempt and #641 for a revived one; the latter PR contains a lot of discussion, including requests from the maintainers to consolidate the work floating around in various branches into a single PR that could be looked at.

Mandatory disclaimer: I know basically nothing about the shim code base, so what I'm doing here consists in 1) taking code other people have written followed by 2) doing whatever is needed to make the compiler stop shouting at me. I am under no delusion that any of this would clear the bar for merging as is, hence the draft status. It's intended to be primarily a conversation starter. With the maintainers' guidance, hopefully we can refine it into something suitable for upstream.

I will have a second PR building on this one soon, which will take care of actually adding riscv64 support.

jmontleon committed:
Make string function argument types match gnu-efi's requirements.

Upstream gnu-efi takes the point of view that string function arguments
in EFI should match the requirements set by the processor's ELF psABI,
even though these aren't ELF.

This patch makes those arguments conform to the resulting types.

Signed-off-by: Jason Montleon <jmontleo@redhat.com>

jmontleon committed:
Correct signedness when calling string functions.

Signed-off-by: Jason Montleon <jmontleo@redhat.com>

gmbr3 committed:
Remove GNU_EFI_USE_EXTERNAL_STDARG

Recent versions of gnu-efi have chosen to provide different API
functions to provide the different ABIs for variadic functions, i.e.
MS_VPoolPrint for the msabi version of VPoolPrint, rather than defining
compile time defines to automatically get compatibility.

As a result, defining GNU_EFI_USE_EXTERNAL_STDARG is not the preferred
way to access these functions.

Signed-off-by: Callum Farmer <gmbr3@opensuse.org>

gmbr3 committed:
Remove GNU_EFI_USE_EXTERNAL_STDARG

Signed-off-by: Callum Farmer <gmbr3@opensuse.org>

gmbr3 committed:
Use ABI-specific variadic print functions

Recent versions of gnu-efi have chosen to provide different API
functions to provide the different ABIs for variadic functions, i.e.
MS_VPoolPrint for the msabi version of VPoolPrint, rather than defining
compile time defines to automatically get compatibility.

This patch uses those functions directly.

Signed-off-by: Callum Farmer <gmbr3@opensuse.org>

gmbr3 committed:
Avoid misuse of Print sys_va_list funcs

Signed-off-by: Callum Farmer <gmbr3@opensuse.org>

gmbr3 committed:
Make: use TOPDIR with gnu-efi paths

Signed-off-by: Callum Farmer <gmbr3@opensuse.org>

gmbr3 committed:
fix gnu-efi paths

Signed-off-by: Callum Farmer <gmbr3@opensuse.org>

gmbr3 committed:
avoid conflicting CompareGuid

Signed-off-by: Callum Farmer <gmbr3@opensuse.org>

gmbr3 committed:
Avoid conflicting CompareGuid

CompareGuid is now defined at two places, this patch keeps them from
breaking the build.

Signed-off-by: Callum Farmer <gmbr3@opensuse.org>

gmbr3 committed:
disable gnuefi_signed_strncmp

Signed-off-by: Callum Farmer <gmbr3@opensuse.org>

andreabolognani committed:
Adopt modern ReallocatePool() ABI

We could theoretically set GNU_EFI_USE_REALLOCATEPOOL_ABI=0
to keep using the legacy ABI, but since gnu-efi uses the
modern ABI internally and we call into its build systemd
directly, doing that messes things up.

Switching to the new ABI is just a matter of changing the
order of arguments.

Signed-off-by: Andrea Bolognani <abologna@redhat.com>

andreabolognani committed:
Adopt modern ReallocatePool() API

Recent versions of gnu-efi swapped the arguments for ReallocPool(),
requiring a special define to get the prior definitions.

We could theoretically set GNU_EFI_USE_REALLOCATEPOOL_ABI=0
to keep using the legacy ABI, but since gnu-efi uses the
modern ABI internally and we call into its build system
directly, doing that messes things up.

Switching to the new ABI is just a matter of changing the
order of arguments.

Signed-off-by: Andrea Bolognani <abologna@redhat.com>

andreabolognani committed:
Set NO_GLIBC=1 when building gnu-efi

shim is a standalone EFI application so it shouldn't be
necessary to look at the glibc headers when building it, and
in fact attempting to do so results in a build failure.

Signed-off-by: Andrea Bolognani <abologna@redhat.com>

andreabolognani committed:
Set NO_GLIBC=1 when building gnu-efi

In recent gnu-efi versions, NO_GLIBC=1 is required to build standalone
applications without glibc headers.

As shim is a standalone EFI application, it shouldn't be
necessary to look at the glibc headers when building it, and
in fact attempting to do so results in a build failure.

Signed-off-by: Andrea Bolognani <abologna@redhat.com>

andreabolognani committed:
gnu-efi: Switch to upstream 4.0.4 release

Replace the shim-specific fork with the upstream version,
specifically the most recent release. Some adjustment to
shim's code are necessary to adapt to this change.

Signed-off-by: Andrea Bolognani <abologna@redhat.com>

andreabolognani committed:
gnu-efi: Switch to upstream 4.0.4 release

Replace the shim-specific fork with the upstream version,
specifically the most recent release. Some adjustment to
shim's code are necessary to adapt to this change.

Signed-off-by: Andrea Bolognani <abologna@redhat.com>

andreabolognani commented:
Tagging @gmbr3 and @jmontleon since some of the changes come from them.

gmbr3 commented:
I'll try and rework some of my commits (which obviously were purely test and look badly formatted early next week) - I should be less busy now since I've left university

gmbr3 committed:
Fix some lds issues

* De-duplicate uses of .note.gnu.build-id/.eh_frame
* Push .reloc after .data (see
  ncroxon/gnu-efi@03bfe2f5ced3cfda8cc693bbb2ff647b8087d5a8)
* ARM updates:
  * use new .text placement (0x1000) fixes allocation issue
    (ncroxon/gnu-efi@24a4cd0e5653fd84b004c00c808c45cc3fb7a7e2)
  * add needed symbols from gnu-efi
  * Add missed reloc section
    (ncroxon/gnu-efi@eadee98bf007a577d461cf9529eae655c9aebd03)

Signed-off-by: Callum Farmer <gmbr3@opensuse.org>

gmbr3 committed:
Fix some lds issues

* De-duplicate uses of .note.gnu.build-id/.eh_frame
* Push .reloc after .data (see
  ncroxon/gnu-efi@03bfe2f5ced3cfda8cc693bbb2ff647b8087d5a8)
* ARM updates:
  * use new .text placement (0x1000) fixes allocation issue
    (ncroxon/gnu-efi@24a4cd0e5653fd84b004c00c808c45cc3fb7a7e2)
  * add needed symbols from gnu-efi
  * Add missed reloc section
    (ncroxon/gnu-efi@eadee98bf007a577d461cf9529eae655c9aebd03)

Signed-off-by: Callum Farmer <gmbr3@opensuse.org>

andreabolognani commented:
Updating now that https://github.com/ncroxon/gnu-efi/pull/82 has been merged and gnu-efi 4.0.4 has been tagged.

With all that in place, I was able to undo the one remaining revert, meaning that a shim binary built from this branch will have feature parity with one built from the main branch.

The git history is still a bit of a mess and clearly unfit for merging as is, but I feel that this is the closest we've ever been to finally making this critical step towards riscv64 support!

@gmbr3 can you please go through your changes again and polish them up as needed?

@vathpela it would be great if you could take a look too. Even in its current state, it should be possible to do at least a first-pass review of the code and provide some high-level feedback that can guide further work.

Thank you both in advance!

andreabolognani commented:
Rebased to match the current contents of the `main` branch.

vathpela commented:
I don't understand what's going on with the `MS_`-prefixed string functions here.  It seems to me that either we're getting compiled with `-mabi=ms_abi` or `-mabi=sysv_abi`, and the exposed API should be the same either way.  What's the point of these wrappers?

/cc @gmbr3

vathpela commented:
And please post a version with the DCI checks fixed - the main CI doesn't run without those, and that is important to have done *before* evaluating patches.

andreabolognani commented:
Picked up a couple additional commits from @gmbr3's [`gnu_efi`](https://github.com/gmbr3/shim/tree/gnu_efi) branch, plus some other minor tweaks. With these changes, both the native test cases and the cross builds should work fine. At least that's the case when trying stuff out locally, let's see what the actual CI thinks about it :)

Callum, there's still a question from @vathpela [above](https://github.com/rhboot/shim/pull/777#issuecomment-4092224609) for you to answer, and I assume you will also want to polish some of the commit messages. I hope you don't mind that I grabbed stuff from your WIP branch, I'm just trying to get things to a place where they can undergo at least some preliminary review. I'm happy to work with you to integrate things better, just let me know how you want to go about it!

andreabolognani committed:
Adopt modern CompareGuid() ABI

Recent versions of gnu-efi change the return semantics of CompareGuid,
requiring defining GNU_EFI_USE_COMPARE_ABI=0 to get the old interface.

We could theoretically set GNU_EFI_USE_COMPARE_ABI=0 to keep using the
legacy ABI, but since gnu-efi uses the modern ABI internally and we call
into its build system directly, doing that messes things up.

In a very small handful of cases we actually rely on the behavior of the
old ABI because we don't just need to know whether or not the two GUIDs are
identical, but also their relative sorting order. CompareGuidForSorting(),
which retains the old behavior, is introduced to deal with those scenarios.

Signed-off-by: Andrea Bolognani <abologna@redhat.com>

andreabolognani committed:
Adopt modern CompareGuid() ABI

We could theoretically set GNU_EFI_USE_COMPARE_ABI=0 to keep
using the legacy ABI, but since gnu-efi uses the modern ABI
internally and we call into its build systemd directly, doing
that messes things up.

In a very small handful of cases we actually rely on the
behavior of the old ABI because we don't just need to know
whether or not the two GUIDs are identical, but also their
relative sorting order. CompareGuidForSorting(), which retains
the old behavior, is introduced to deal with those scenarios.

Signed-off-by: Andrea Bolognani <abologna@redhat.com>

andreabolognani commented:
Rebased on top of `main` and adopted the modern `CompareGuid` API/ABI. The `mkosi` tests should pass now.

andreabolognani commented:
Picked another commit off @gmbr3's `gnu_efi` branch, with some tweaks so that it applies. The arm cross-build now succeeds locally.

andreabolognani commented:
Marking as ready to review since CI passes now.

@gmbr3 I hope that you will be able to spend some time on this. There are some questions from @vathpela up above that I am unfortunately unable to answer myself, and possibly you'll want to extend and polish the commit messages a bit.

We are _incredibly_ close to finally getting this (and riscv64 support) merged. Let's get it over the finish line!

gmbr3 commented:
> I don't understand what's going on with the `MS_`-prefixed string functions here. It seems to me that either we're getting compiled with `-mabi=ms_abi` or `-mabi=sysv_abi`, and the exposed API should be the same either way. What's the point of these wrappers?
> 
> /cc @gmbr3

gnu-efi by default compiles using SysV va_list ABI (we don't specify any -mabi), shim in some places needs MS va_list ABI. 

These prefixed functions were created to allow building both in one library

The exposed API differs in the internal types va_list on SysV and char* on MS which causes a build failure

brianredbeard commented on .gitmodules:
Is this supposed to be master or v4.0.4?  If this is a problem the same applies to #778

andreabolognani replied:
The branch is master, the specific commit is https://github.com/ncroxon/gnu-efi/commit/37cd8f069bde6715eebdc5e38a8f15ee6de5edcf i.e. the [4.0.4 tag](https://github.com/ncroxon/gnu-efi/releases/tag/4.0.4). If switching to another release will be desired in the future, it will be enough to change the specific commit the submodule is pointing to without having to touch the `.gitmodules` file again.

vathpela committed:
test: Remove the test for gnuefi_signed_strncmp.

We don't need this at all any more.

Signed-off-by: Peter Jones <pjones@redhat.com>
